lib/core/base/rule.js

```diff
@@ -242,20 +244,131 @@ Rule.prototype.run = function(context, options, resolve, reject) {
 	q.defer(resolve => setTimeout(resolve, 0));
 
 	if (options.performanceTimer) {
-		axe.utils.performanceTimer.mark(markChecksEnd);
-		axe.utils.performanceTimer.mark(markEnd);
-		axe.utils.performanceTimer.measure(
-			'rule_' + this.id + '#runchecks',
-			markChecksStart,
-			markChecksEnd
-		);
-
-		axe.utils.performanceTimer.measure('rule_' + this.id, markStart, markEnd);
+		logRulePerformance(this);
 	}
 
 	q.then(() => resolve(ruleResult)).catch(error => reject(error));
 };
 
+/**
+ * Runs the Rule's `evaluate` function synchronously
+ * @param  {Context}   context  The resolved Context object
+ * @param  {Mixed}   options  Options specific to this rule
+ */
+Rule.prototype.runSync = function(context, options = {}) {
+	if (options.performanceTimer) {
+		this._trackPerformance();
+	}
+
+	const ruleResult = new RuleResult(this);
+	let nodes;
+
+	try {
+		nodes = this.gatherAndMatchNodes(context, options);
+	} catch (error) {
+		// Exit the rule execution if matches fails
+		throw new SupportError({ cause: error, ruleId: this.id });
+	}
+
+	if (options.performanceTimer) {
+		logGatherPerformance(this, nodes);
+	}
+
+	nodes.forEach(node => {
+		let results = [];
+		['any', 'all', 'none'].forEach(type => {
+			results.push(this.runChecksSync(type, node, options, context));
+		});
+
+		const result = getResult(results);
+		if (result) {
+			result.node = new axe.utils.DqElement(node.actualNode, options);
+			ruleResult.nodes.push(result);
+		}
+	});
+
+	if (options.performanceTimer) {
+		logRulePerformance(this);
+	}
+
+	return ruleResult;
+};
+
+/**
+ * Add performance tracking properties to the rule
+ * @private
+ */
+Rule.prototype._trackPerformance = function() {
+	this._markStart = 'mark_rule_start_' + this.id;
+	this._markEnd = 'mark_rule_end_' + this.id;
+	this._markChecksStart = 'mark_runchecks_start_' + this.id;
+	this._markChecksEnd = 'mark_runchecks_end_' + this.id;
+};
+
+/**
+ * Log performance of rule.gather
+ * @param {Rule} rule The rule to log
+ * @param {Array} nodes Result of rule.gather
+ */
+function logGatherPerformance(rule, nodes) {
```

Comment:
I think all these new functions should be methods on the Rule class. That makes the implementation consistent.